File: Android/AnyMind/app/src/main/java/com/anymind/anymind/sync/SyncClient.kt
package com.anymind.anymind.sync

import com.google.gson.Gson
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.ResponseBody

class SyncClient(
    private val client: OkHttpClient = OkHttpClient(),
    private val gson: Gson = Gson()
) {
    fun push(baseUrl: String, request: SyncPushRequest): SyncPushResponse {
        return post(baseUrl, "/sync/push", request, SyncPushResponse::class.java)
    }

    fun pull(baseUrl: String, request: SyncPullRequest): SyncPullResponse {
        return post(baseUrl, "/sync/pull", request, SyncPullResponse::class.java)
    }

    fun createSpace(baseUrl: String, name: String?): SpaceCreateResponse {
        return post(baseUrl, "/spaces", SpaceCreateRequest(name), SpaceCreateResponse::class.java)
    }

    private fun <T : Any, R : Any> post(
        baseUrl: String,
        path: String,
        body: T,
        responseType: Class<R>
    ): R {
        val url = baseUrl.trimEnd('/') + path
        val json = gson.toJson(body)
        val requestBody = json.toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url(url)
            .post(requestBody)
            .build()
        val response = client.newCall(request).execute()
        response.use {
            if (!response.isSuccessful) {
                throw SyncException("HTTP ${response.code}")
            }
            val responseBody = response.body ?: throw SyncException("Empty response")
            return parse(responseBody, responseType)
        }
    }

    private fun <R : Any> parse(body: ResponseBody, responseType: Class<R>): R {
        val text = body.string()
        return gson.fromJson(text, responseType)
    }
}

class SyncException(message: String) : RuntimeException(message)

File: ios/AnyMind/PromptRecorder/Sync/SyncClient.swift
import Foundation

final class SyncClient {
    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    func push(baseURL: URL, request: SyncPushRequest) async throws -> SyncPushResponse {
        try await post(path: "/sync/push", baseURL: baseURL, body: request)
    }

    func pull(baseURL: URL, request: SyncPullRequest) async throws -> SyncPullResponse {
        try await post(path: "/sync/pull", baseURL: baseURL, body: request)
    }

    func createSpace(baseURL: URL, name: String?) async throws -> SpaceCreateResponse {
        try await post(path: "/spaces", baseURL: baseURL, body: SpaceCreateRequest(name: name))
    }

    private func post<T: Codable, R: Codable>(path: String, baseURL: URL, body: T) async throws -> R {
        let url = baseURL.appendingPathComponent(path)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await session.data(for: request)
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            throw SyncError.httpStatus(http.statusCode)
        }
        return try JSONDecoder().decode(R.self, from: data)
    }
}

enum SyncError: Error {
    case missingConfig
    case httpStatus(Int)
}
